### DISCUSSION TITLE
How to handle `LAMBDA_RUNTIME Failed to post handler success response` in Next.js

### DISCUSSION THREAD

**Discussion by adamhenson** [vercel/vercel#9151] [Help]
I have a Next.js app with API routes and it is deployed on Vercel. My issue is that I don't know how to handle 500 types of errors in the Next.js app that occur within the Lambda runtime. The effect is that my API endpoint responds with HTML content (the default Next.js 500 page). I can only reproduce this in Vercel... not locally - due to specific runtime in Vercel.

This is a screenshot of my issue.

<img width="1405" alt="Screen Shot 2022-12-31 at 11 41 56 AM" src="https://user-images.githubusercontent.com/747743/210150223-92c2ef0f-563f-493d-a205-baed477bda67.png">

So, the failure message my frontend code receives is really an issue in parsing HTML content from the API endpoint when it expects JSON. `Unexpected token '<', "<!DOCTYPE "... is not valid JSON`. So, even with the try / catch block, it seems that the error occurs outside of the scope of my handler, bypassing any control I have over it (seemingly).

This is my API handler code:

```typescript
export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse<ResponseData>,
) {
  try {
    const data = {/* something very large :/ */}
    // some obviously problematic code that generates a ginourmous payload... exceeding 4mb
    return res.status(200).json({ data });
  } catch (error: any) {
    return res
      .status(500)
      .json({ error: error?.message || 'something went wrong' });
  }
}
```

**Expected**

A JSON response with a 500 status code that includes an intuitive message like so (or even better a 413 status code as reflected in the Vercel logs).

```json
{
  "error": "exceeds 4MB. API Routes are meant to respond quickly."
}
```

**Actual**

A generic 500 status code HTML page response (the Next.js default)

**Question**

Can I somehow handle this error and control the response? I'm thinking of the Express.js global error handling from errors that bubble all the way up. I looked into a way of handling this in Next.js middleware, but because this issue occurs in the response, not the request... seems like it wouldn't the right path for a solution.

**Comment by IgorKowalczyk** [comment]
You can check out https://vercel.com/docs/concepts/functions/edge-functions#limitations

**Reply by adamhenson** [reply]
Thanks @IgorKowalczyk. That is interesting, however not a specific answer I'm looking for. But, perhaps there is no answer and this is simply a limitation. Since API routes run at the edge, there is not a way to handle the 500 error that is surfaced on the edge within the Next.js app. However, somehow Next.js knows to respond with its own 500 status page.

My question that I'm not able to find an answer to is - how can I handle this, so that the result isn't a response of the default HTML Next.js 500 status page? Can I somehow create an API route specific 500 page, so at least the response is in JSON format?

**Reply by adamhenson** [reply]
Maybe this would be a good feature to request, if I'm understanding everything correctly - an ability to create an edge error handler?

**Reply by IgorKowalczyk** [reply]
If I remember everything correctly - you can't create your own error handler on the edge. Edge is a very limited environment, not all features of normal routes are available

**Reply by adamhenson** [reply]
Thanks @IgorKowalczyk. Yes, maybe this is all over my head, because I would think since the API route handlers run at the edge and this error is thrown at the edge then this error should be caught within the API route handler, so it could be and handled by it. But, there seems to be something that isn't clicking for me.

Anyways, even with this limitation, I think it would be awesome if the mechanism that catches the error and serves the default Next.js HTML 500 error would instead be smart enough to know this API route is supposed to respond with JSON content. Or maybe some way to support a user defined API route specific 500 page.

I'm probably asking for too much since I realize the these are brand new features. It seems like for now, I'll just need to deal with this limitation, which isn't the worst, but a REST API that can sometimes respond with HTML content just doesn't sit well in my idealist head :)

Regardless, I'm grateful for these new features and very excited to see how they will evolve.
